### DISCUSSION TITLE
Mail template translations not working

### DISCUSSION THREAD

**Discussion by AIC-BV** [wintercms/winter#327] [Q&A]
Hello!

I am trying to translate my Mail Templates, which is working by adding the following code in my boot() function:
>         \System\Models\MailTemplate::extend(function ($model) {
>             $model->addDynamicProperty('translatable', []);
>             $model->translatable = array_merge($model->translatable, ['subject@create', 'subject@update', 'description', 'content_html', 'content_text']);
>             $model->extendClassWith('Winter\Translate\Behaviors\TranslatableModel');
>         });

Just one small issue, even though the mail is translatable in the backend at `settings -> mail templates`, the mail always sends in the default language and doesn't listen to the activeLocale (I'm using Winter.Translate).

Any ideas how to solve this problem so that the mail sends in the correct language?
I am sending the mail with a queue:

        Mail::queue(aic.contact::mail.contact, $data, function($message) use (&$data) {
            // get the theme variables
            $sender = $theme->mail_email;
            $senderName = $theme->mail_name;

            // receiver
            $email = $data['email'];
            $name = $data['fname'] . ' ' . $data['lname'];

            // set properties
            $message->from($sender, $senderName);
            $message->to($email, $name);
        });

Edit:
The suggested solution from daftspunk, which was already implemented in WinterCMS, does not seem to solve my issue: https://github.com/wintercms/storm/pull/50

Edit 2:
Maybe related to this issue;
I have this field in 'fields.yaml' for my theme:

```
  street:
        tab: Contact
        label: Straat
        type: text
        span: left
        translatable: true
```

When printing my theme variable: `{{ this.theme.street }}`; it's always in the default language and my translated value is ignored. I'm absolutely sure this used to work a few days ago. Maybe I broke something? I didn't touch any plugins and commenting my plugin code doesn't affect this issue, and the translations are visible in the database in `winter_translate_attributes`

**Comment by bennothommo** [comment]
@AIC-BV what happens when you include the current locale in `$data` provided to the `Mail::queue` call?

```
$data['_current_locale'] = App::getLocale();
```

My guess is that when `Winter\Storm\Mail\Mailable` is rendered on the queued job being run, the `App::getLocale()` that fires there would be using the default locale because the job runner doesn't have a user session, so it wouldn't be able to get the *user's* locale.

**Reply by AIC-BV** [reply]
Thank you for time!
Good guess! But sadly it doesn't work and the mail is still in the default langauge

**Reply by AIC-BV** [reply]
I found a bug, perhaps related to this issue.

As stated in my 'edit 2' part of the original post;
> When printing my theme variable: {{ this.theme.street }}; it's always in the default language and my translated value is ignored.

I noticed that this is because of `View::share`.

For example `View::share('website_logo',  $theme->website_logo);`
When I comment the `View::share` code, the theme translations start working again.


Submitted a bug report for this issue: [https://github.com/wintercms/winter/issues/330](https://github.com/wintercms/winter/issues/330)

The original issue is still unsolved.
I was thinking about creating a template for each language, to work around this issue...: `contact::en`, `contact::fr` , ... and then sending the correct template: `$template = 'contact::' . App:getLocale();`, but this is more complicated for the client

**Reply by bennothommo** [reply]
Hmmm, you're right.

We'll have to look into this further, in the mean time, you could try this plugin: https://github.com/vdomah/oc-mailtranslatable-plugin

**Reply by AIC-BV** [reply]
I changed `queue` to `send`. This solved the problem.
Do you think its a bad idea? Should mails always be send with a queue to prevent server load?

If so, I think i'll just stay to my suggested solution by making a template for each language and use it with `$template = 'contact::' . App:getLocale()`.

I prefer to avoid 3rd party plugins and write my own stuff, which I understand and can debug if needed. (with the exception of the awesome Winter Plugins ofcourse!)

**Reply by bennothommo** [reply]
Queued emails are generally a good idea, but if you're not sending too many at a time, then using `send` is perfectly fine.
